File: java/src/com/chargebee/samples/WebhookHandler.java
/*
 * Copyright (c) 2013 ChargeBee Inc
 * All Rights Reserved.
 */
package com.chargebee.samples;

import com.chargebee.Result;
import com.chargebee.models.Event;
import com.chargebee.models.Invoice;
import com.chargebee.models.enums.EventType;
import com.chargebee.org.json.JSONException;
import com.chargebee.org.json.JSONObject;
import java.beans.EventSetDescriptor;
import java.io.BufferedReader;
import java.io.IOException;
import java.util.logging.Logger;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

/*
 * Demo on how to add charge for meter billing customer after
 * receiving Invoice Created event through webhook.
 */
public class WebhookHandler extends HttpServlet {

    
    /*
     * Receives the webhook content from ChargeBee.
     */
    @Override
    protected void doPost(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        
        if(!checkIfRequestIsFromChargeBee(request, response)){
            return;
        }
        
        /*
         * Getting the json content from the request.
         */
        BufferedReader reader = request.getReader();

        /* 
         * Assigning the recieved content to ChargeBee Event object.
         */
        Event event = new Event(reader);
        

        
        /*
         * Checking the event type as Invoice Created to add Charge for Meter Billing.
         */
        EventType eventType = event.eventType();
        if (EventType.INVOICE_CREATED.equals(eventType)) {
            new MeterBilling().closePendingInvoice(event.content().invoice());
        }
        
    }

    
    
    /**Check if the request is from chargebee. 
     * You can secure the webhook either using
     *   - Basic Authentication
     *   - Or check for specific value in a parameter.
     *<br/>
     * For demo purpose we are using the second option though basic auth is strongly
     * preferred. Also store the key securely in the server rather than hard coding in code.
     */
    private static boolean checkIfRequestIsFromChargeBee(HttpServletRequest req,
            HttpServletResponse resp) throws IOException{
        if(!"DEMO_KEY".equals(req.getParameter("webhook_key"))){
            resp.sendError(HttpServletResponse.SC_FORBIDDEN,"webhook_key not correct");
            return false;
        }
        return true;
    }
    
    
    /**
     * Returns a short description of the servlet.
     *
     * @return a String containing servlet description
     */
    @Override
    public String getServletInfo() {
        return "Demo on how to add charge for meter billing customer after " +
                "receiving Invoice Created event through webhook.";
    }
}

File: php/php_src/WebhookHandler.php
<?php
/*
 * Adding ChargeBee php libraries and configuration files.
 */
require_once(dirname(__FILE__) . "/Config.php");
include("./MeterBilling.php");
/*
 * Demo on how to add charge for meter billing customer after
 * receiving Invoice Created event through webhook.
 */
if (isset($_POST)) {
  
  /*
   * Getting the json content from the request.
   */
  if(!checkIfRequestIsFromChargeBee() ){
    return;
  }
  $content = file_get_contents('php://input');
  /* 
   * Assigning the recieved content to ChargeBee Event object.
   */
  $event = ChargeBee_Event::deserialize($content);
  

  
  /*
   * Checking the event type as Invoice Created to add Charge for Meter Billing.
   */
  $eventType = $event->eventType;
  if($eventType == "invoice_created" ) {
	 $invoiceId = $event->content()->invoice()->id;
     $invoiceObj = ChargeBee_Invoice::retrieve($invoiceId)->invoice();
	 if($invoiceObj->status == "pending" ){
       $meterBilling = new MeterBilling();
       $meterBilling->closePendingInvoice($invoiceObj);
	   echo "Invoice has been closed successfully";
     }else {
		 echo "Invoice is not in pending state";
     }
  }
  

}


/* Check if the request is from chargebee. 
 * You can secure the webhook either using
 *   - Basic Authentication
 *   - Or check for specific value in a parameter.
 * For demo purpose we are using the second option though 
 * basic auth is strongly preferred. Also store the key 
 * securely in the server rather than hard coding in code.
 */
function checkIfRequestIsFromChargeBee() {
 if($_REQUEST['webhook_key'] != "DEMO_KEY" ) { 
     header("HTTP/1.0 403 Error");
     return false;
 }
 return true;
}


?>
